### PR TITLE
sonarlint-ls: init at 3.5.1.75119

### PR DESCRIPTION
**Pull request by tricktron**
## Description of changes

The [Sonarlint language server](https://github.com/SonarSource/sonarlint-language-server) builds the basis for the [sonarlint neovim plugin](https://gitlab.com/schrieveslaach/sonarlint.nvim) and the [Sonarlint VS code extension](https://marketplace.visualstudio.com/items?itemName=SonarSource.sonarlint-vscode).
 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/so/sonarlint-ls/package.nix [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+, writeScript
+, runCommand
+, sonarlint-ls
+}:
+
+let mavenJdk17 = maven.override { jdk = jdk17; };
+
+in
+
+mavenJdk17.buildMavenPackage rec {
```
Comment (pluiedev) [inline pkgs/by-name/so/sonarlint-ls/package.nix:16]: Style nit
```suggestion
let 
  mavenJdk17 = maven.override { jdk = jdk17; };
in
mavenJdk17.buildMavenPackage rec {
```

File: pkgs/by-name/so/sonarlint-ls/package.nix [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    !LanguageServerWithFoldersMediumTests,
+    !NotebookMediumTests,
+    !ConnectedModeMediumTests,
+    !JavaMediumTests"
+  ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin $out/share $out/share/plugins
```
Comment (pluiedev) [inline pkgs/by-name/so/sonarlint-ls/package.nix:55]:
```suggestion
    mkdir -p $out/{bin,share/plugins}
```

File: pkgs/by-name/so/sonarlint-ls/package.nix [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      sonarlint-ls-starts-successfully = runCommand "${pname}-test" {} ''
+        ${sonarlint-ls}/bin/sonarlint-ls > $out
+        cat $out | grep "SonarLint backend started"
+      '';
+    };
+
+    updateScript =
+    let pkgFile = builtins.toString ./package.nix;
+    in
+    writeScript "update-${pname}" ''
```
Comment (pluiedev) [inline pkgs/by-name/so/sonarlint-ls/package.nix:83]:
```suggestion
    updateScript = let 
      pkgFile = builtins.toString ./package.nix;
    in
    writeScript "update-${pname}" ''
```

File: pkgs/by-name/so/sonarlint-ls/package.nix [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+      NEW_MVN_HASH=$(cat ${pname}-stderr.log | grep "got:" | awk '{print ''$2}')
+      rm ${pname}-stderr.log
+      # escaping double quotes looks ugly but is needed for variable substitution
+      # use # instead of / as separator because the sha256 might contain the / character
+      sed -i "0,/mvnHash *= *\"[^\"]*\"/{s#mvnHash = \"[^\"]*\"#mvnHash = \"$NEW_MVN_HASH\"#}" ${pkgFile}
+    '';
+  };
+
+  meta = with lib; {
```
Comment (pluiedev) [inline pkgs/by-name/so/sonarlint-ls/package.nix:105]: [Avoid `with lib;`](https://github.com/NixOS/nixpkgs/issues/208242)

File: pkgs/by-name/so/sonarlint-ls/package.nix [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        ${sonarlint-ls}/bin/sonarlint-ls > $out
+        cat $out | grep "SonarLint backend started"
+      '';
+    };
+
+    updateScript =
+    let pkgFile = builtins.toString ./package.nix;
+    in
+    writeScript "update-${pname}" ''
+      #!/usr/bin/env nix-shell
```
Comment (pluiedev) [inline pkgs/by-name/so/sonarlint-ls/package.nix:84]: Maybe consider using `writeShellApplication` instead?
```nix
lib.getExe (writeShellApplication {
  name = "...";
  runtimeInputs = [curl pcre common-updater-scripts jq gnused];
  text = ''
    LATEST_TAG = ...
    #...
  '';
})
```

**Review by pluiedev — CHANGES_REQUESTED** [review]

File: pkgs/by-name/so/sonarlint-ls/package.nix (hunk 4 above)
Comment (tricktron) [reply pkgs/by-name/so/sonarlint-ls/package.nix]: Done 👍

File: pkgs/by-name/so/sonarlint-ls/package.nix (hunk 5 above)
Comment (tricktron) [reply pkgs/by-name/so/sonarlint-ls/package.nix]: Done 👍

**Comment by tricktron** [comment]
@pluiedev Thank you for the quick review.  I applied all your suggestions.

**Review by pluiedev — APPROVED** [review]
LGTM, but please squash all commits into just one before merging.

**Comment by tricktron** [comment]
@pluiedev Thanks. I have a question regarding the best practice for squashing commits: Won´t the commits be squashed automatically anyway when merging? So what is the benefit of squashing them again before the merge?

I am asking because I quite like having the full history (single commits) in the pull request because it happened more than once that I remembered something I have done for another package, then got confused why I have done thing X, click on my pull request and finally realise thanks to the whole history that thing X was replaced by best practice Y :smiley: .

**Comment by pluiedev** [comment]
> @pluiedev Thanks. I have a question regarding the best practice for squashing commits: Won´t the commits be squashed automatically anyway when merging? So what is the benefit of squashing them again before the merge?
> 
> I am asking because I quite like having the full history (single commits) in the pull request because it happened more than once that I remembered something I have done for another package, then got confused why I have done thing X, click on my pull request and finally realise thanks to the whole history that thing X was replaced by best practice Y 😃 .

Nixpkgs committers usually don't do squash merges :sweat_smile: That's why you see people amending and force pushing stuff all the time, or squashing them before merging

**Review by pluiedev — APPROVED** [review]

**Pull request merged**